From d754dd50c4eaee53bcfad0eb83dc4fe54e741252 Mon Sep 17 00:00:00 2001
From: Bo Huang <bo.b.huang@intel.com>
Date: Fri, 24 May 2013 17:14:09 +0800
Subject: MTP server: MTP Server Tombstone after unplug/plug USB cable during transfering file.

BZ: 107499

Rootcause:
When user unplug USB cable during transfering file from PC to DUT,
MTP server call unlink system function to remove dirty file, and then call implemenation
of MTP java to update object information. At same time, USB driver notify MTP service in
Java space with USB disconnect event. MTP service release instance of MTP server and
MTP database. If unlink function block MTP server, MTP server will tombstone due to illegal access.
Solution:
Move unlink call for EIO error case.

Category: AOSP Improvement
Domain: USB MTP-PTP
Origin: Internal
Upstream-Candidate: No, proprietary

Change-Id: Ib1581f378e4e63277ef3099f278222390e7fff09
Orig-Change-Id: Ie3a8037ef85f4ccbd2b890a453f112788a73047b
Signed-off-by: Bo Huang <bo.b.huang@intel.com>
---
 media/mtp/MtpServer.cpp |   12 +++++++++++-
 media/mtp/MtpServer.h   |    1 +
 2 files changed, 12 insertions(+), 1 deletions(-)

diff --git a/media/mtp/MtpServer.cpp b/media/mtp/MtpServer.cpp
index d5f76ae..9dc9500 100644
--- a/media/mtp/MtpServer.cpp
+++ b/media/mtp/MtpServer.cpp
@@ -235,6 +235,11 @@ void MtpServer::run() {
         mDatabase->sessionEnded();
     close(fd);
     mFD = -1;
+
+    if (mDirtyFilePath.length() != 0) {
+        ALOGD("Delete dirty file %s. \n", (const char *)mDirtyFilePath);
+        unlink(mDirtyFilePath);
+    }
 }
 
 void MtpServer::sendObjectAdded(MtpObjectHandle handle) {
@@ -959,7 +964,12 @@ MtpResponseCode MtpServer::doSendObject() {
     close(mfr.fd);
 
     if (ret < 0) {
-        unlink(mSendObjectFilePath);
+        if (errno == EIO) {
+            mDirtyFilePath = mSendObjectFilePath;
+        } else {
+            unlink(mSendObjectFilePath);
+        }
+
         if (errno == ECANCELED || errno == EIO )
             result = MTP_RESPONSE_TRANSACTION_CANCELLED;
         else
diff --git a/media/mtp/MtpServer.h b/media/mtp/MtpServer.h
index dfa8258..bdcdf42 100644
--- a/media/mtp/MtpServer.h
+++ b/media/mtp/MtpServer.h
@@ -64,6 +64,7 @@ private:
     MtpObjectHandle     mSendObjectHandle;
     MtpObjectFormat     mSendObjectFormat;
     MtpString           mSendObjectFilePath;
+    MtpString           mDirtyFilePath;
     size_t              mSendObjectFileSize;
 
     Mutex               mMutex;
-- 
1.7.4.1

